llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2114,13 +2129,16 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
     if (IntrinsicID == Intrinsic::canonicalize)
       return constantFoldCanonicalize(Ty, Call, U);
 
-#if defined(HAS_IEE754_FLOAT128) && defined(HAS_LOGF128)
+#if defined(HAS_IEE754_FLOAT128)
     if (Ty->isFP128Ty()) {
       if (IntrinsicID == Intrinsic::log) {
-        float128 Result = logf128(Op->getValueAPF().convertToQuad());
+        APFloat Value = Op->getValueAPF();
+        if (!Value.isValidIEEEQuad())
+          return nullptr;
+
+        float128 Result = logf128(ConvertToQuad(Value));
```

Comment:
`APFloat` should have a software emulator that behaves bit-identical.
Then, the native `logf128` may be used as an optimization in it.